README.md
clockdrift.js
=============

Node script for detecting clock drift on remote web servers

* Author: Rich Trott
* Copyright: Regents of the University of California (c) 2012
* License: MIT

## What is it?

Shibboleth authentication via simpleSAMLphp broke on one of our apps when the system clock on the Shibboleth server drifted more than a minute into the future. 

I am not the administrator of that machine and I don't know if they fixed whatever was causing the problem. But it occured to me that even with no access to the server, I could detect clock drift by examining the date stamp in HTTP headers from the server.

So, I wrote this tool to use in a crontab.

## How do I install it?

    npm install -g clockdrift

## How do I use it?

    clockdrift 15 http://www.example.com/ https://www.example.net/

The first argument is the drift tolerance in seconds. Clocks that are off from your local clock by less than the specified number of seconds (15 in the example above) will not be reported. Use `-1` to receive notices about all clocks even if they are set precisely.

By the way, since the tool uses your clock as the canonical clock, it's good to take steps to make sure that your system clock isn't the problem. I include the Naval Observatory clock URL at http://tycho.usno.navy.mil/ in my cron job to make sure that it's not my system clock that is off as opposed to the remote server clock.

The remaining arguments are the URLs that the tool should connect to in order to retrieve timestamps. In the example above, we are checking the clocks on `www.example.com` and `www.example.net`.

The tool assumes that you have a fast connection and/or are on a network that is nearby the servers you are checking. Therefore, a one second timeout is hardcoded. This insures against false positives due to slow network connections.

CHANGELOG.md
# [5.0.0](https://github.com/Trott/clockdrift.js/compare/v4.0.1...v5.0.0) (2023-06-15)


### chore

* drop support for Node.js < 16.x ([31b82e9](https://github.com/Trott/clockdrift.js/commit/31b82e998d6dac5f7ee4e4bfcce01eac1c24a305))


### BREAKING CHANGES

* Dropping support for Node.js prior to 16.x

## [4.0.1](https://github.com/Trott/clockdrift.js/compare/v4.0.0...v4.0.1) (2023-02-04)


### Bug Fixes

* require a valid certificate ([7cac352](https://github.com/Trott/clockdrift.js/commit/7cac3521c06da0cad74dffb71ceae3f9b9055ff0))

# [4.0.0](https://github.com/Trott/clockdrift.js/compare/v3.0.0...v4.0.0) (2020-10-26)


### chore

* increase coverage requirement to 100% ([d1e0f08](https://github.com/Trott/clockdrift.js/commit/d1e0f0810e78def1043c1faf221abcdd7513ae0a))


### BREAKING CHANGES

* drop support for Node.js 10.x

# [3.0.0](https://github.com/Trott/clockdrift.js/compare/v2.0.0...v3.0.0) (2020-10-25)


### Bug Fixes

* exit with code when something goes wrong ([#5](https://github.com/Trott/clockdrift.js/issues/5)) ([5cca36a](https://github.com/Trott/clockdrift.js/commit/5cca36a3341926963657a0ac4ebc101f0211cc3c))


### BREAKING CHANGES

* drop support for Node.js 8.x

# [2.0.0](https://github.com/Trott/clockdrift.js/compare/v1.0.0...v2.0.0) (2020-10-25)


### Bug Fixes

* specify correct dependencies ([7af3980](https://github.com/Trott/clockdrift.js/commit/7af3980b567fd338d5939a27d37bc38aa7bfd340))


### chore

* drop support for Node.js prior to 8.0.0 ([#4](https://github.com/Trott/clockdrift.js/issues/4)) ([aed4f43](https://github.com/Trott/clockdrift.js/commit/aed4f43222b8c56c2c1bc1c3bfa9328ca074a2de))


### BREAKING CHANGES

* drop support for Node.js prior to 8.0.0


test.js
'use strict';
import assert from 'node:assert';
import { spawnSync, spawn } from 'node:child_process';
import http from 'node:http';
const usage = 'Usage: clockdrift.js tolerance url1 [url2 ...]\n' +
    '    tolerance: number of seconds a timestamp can be off without being reported\n' +
    '    url1 ...: URLs to inspect\n' +
    '\n' +
    'Example: clockdrift.js 15 http://tycho.usno.navy.mil/ http://www.example.com/\n';
const runSync = (args) => {
    return spawnSync('./clockdrift.js', args, { encoding: 'utf8' });
};
const run = (args, cb) => {
    const subprocess = spawn('./clockdrift.js', args);
    subprocess.on('error', (err) => {
        assert.fail(err);
    });
    subprocess.on('exit', cb);
    return subprocess;
};
{
    // Using no arguments should return the usage message.
    const ret = runSync([]);
    assert.strictEqual(ret.stdout, '');
    assert.strictEqual(ret.stderr, usage);
    assert.strictEqual(ret.status, 1);
    assert.strictEqual(ret.signal, null);
}
{
    // The tolerance argument must be a number.
    const ret = runSync(['fhqwhgads', 'https://www.example.com/']);
    assert.strictEqual(ret.stdout, '');
    assert.strictEqual(ret.stderr, '\nError: tolerance must be a number.\n\n' + usage);
    assert.strictEqual(ret.status, 1);
    assert.strictEqual(ret.signal, null);
}
{
    // Invalid hostnames.
    const ret1 = runSync(['42', 'http://fhqwhgads.invalid/']);
    const ret2 = runSync(['42', 'https://fhqwhgads.invalid/']);
    [ret1, ret2].forEach((ret) => {
        assert.strictEqual(ret.stdout, '');
        assert.ok(ret.stderr.startsWith('Error on fhqwhgads.invalid: getaddrinfo '), ret.stderr);
        assert.strictEqual(ret.status, 1);
        assert.strictEqual(ret.signal, null);
    });
}
{
    // Unrecognized protocol.
    const ret = runSync(['1', 'fhqwhgads://']);
    assert.strictEqual(ret.stdout, '');
    assert.strictEqual(ret.stderr, 'Could not determine protocol: fhqwhgads:\n');
    assert.strictEqual(ret.status, 1);
    assert.strictEqual(ret.signal, null);
}
{
    // Start an http server on localhost and test the happy path.
    const server = http.createServer((req, res) => {
        res.writeHead(200, { Connection: 'close' });
        res.end('fhqwhgads');
    });
    server.listen(0, '127.0.0.1', () => {
        const addressInfo = server.address();
        const ret = run(['1', `http://127.0.0.1:${addressInfo.port}/`], (code, signal) => {
            ret.stdout.on('data', assert.fail);
            ret.stderr.on('data', assert.fail);
            assert.strictEqual(code, 0);
            assert.strictEqual(signal, null);
            server.close();
        });
    });
}
{
    // Test with http server that has an inaccurate Date header.
    const server = http.createServer((req, res) => {
        res.writeHead(200, {
            Date: new Date(Date.now() - 2000).toUTCString(),
            Connection: 'close'
        });
        res.end('fhqwhgads');
    });
    server.listen(0, '127.0.0.1', () => {
        const addressInfo = server.address();
        const port = addressInfo.port;
        const ret = run(['1', `http://127.0.0.1:${port}/`], (code, signal) => {
            ret.stderr.on('data', (msg) => {
                const expected = `Clock at 127.0.0.1:${port} is -2s off from local clock.\n`;
                assert.strictEqual(msg.toString(), expected);
            });
            ret.stderr.on('data', assert.fail);
            assert.strictEqual(code, 0);
            assert.strictEqual(signal, null);
            server.close();
        });
    });
}
{
    // Test with an http server that a malformed Date header.
    const server = http.createServer((req, res) => {
        res.setHeader('Date', 'fhqwhgads');
        res.setHeader('Connection', 'close');
        res.end('fhqwhgads');
    });
    server.listen(0, '127.0.0.1', () => {
        const addressInfo = server.address();
        const port = addressInfo.port;
        const ret = run(['1', `http://127.0.0.1:${port}/`], (code, signal) => {
            ret.stdout.on('data', assert.fail);
            ret.stderr.on('data', (msg) => {
                const expected = `Could not convert date header from 127.0.0.1:${port} to timestamp. (Malformed?)\n`;
                assert.strictEqual(msg.toString(), expected);
            });
            assert.strictEqual(code, 1);
            assert.strictEqual(signal, null);
            server.close();
        });
    });
}
{
    // Test with a broken http server that does not send a Date header.
    const server = http.createServer((req, res) => {
        res.sendDate = false;
        res.writeHead(200, { Connection: 'close' });
        res.end('fhqwhgads');
    });
    server.listen(0, '127.0.0.1', () => {
        const addressInfo = server.address();
        const port = addressInfo.port;
        const ret = run(['1', `http://127.0.0.1:${port}/`], (code, signal) => {
            ret.stdout.on('data', assert.fail);
            ret.stderr.on('data', (msg) => {
                const expected = `No date header returned by 127.0.0.1:${port}.\n`;
                assert.strictEqual(msg.toString(), expected);
            });
            assert.strictEqual(code, 1);
            assert.strictEqual(signal, null);
            server.close();
        });
    });
}
{
    // Test with an http server that times out.
    const server = http.createServer();
    server.listen(0, '127.0.0.1', () => {
        const addressInfo = server.address();
        const port = addressInfo.port;
        const msgs = [
            'Socket timeout; hanging up.',
            `Error on 127.0.0.1:${port}: socket hang up`
        ];
        const ret = run(['1', `http://127.0.0.1:${port}/`], (code, signal) => {
            ret.stdout.on('data', assert.fail);
            ret.stderr.on('data', (chunk) => {
                chunk.toString().split('\n').forEach((msg) => {
                    if (msg.length > 0) {
                        const expected = msgs.shift();
                        assert.strictEqual(msg, expected);
                    }
                });
            });
            ret.stderr.on('end', () => {
                assert.deepStrictEqual(msgs, []);
            });
            assert.strictEqual(code, 1);
            assert.strictEqual(signal, null);
            server.close();
        });
    });
}


test.ts
'use strict'

import assert from 'node:assert'
import { spawnSync, spawn, ChildProcessWithoutNullStreams, SpawnSyncReturns } from 'node:child_process'
import http from 'node:http'

const usage =
  'Usage: clockdrift.js tolerance url1 [url2 ...]\n' +
  '    tolerance: number of seconds a timestamp can be off without being reported\n' +
  '    url1 ...: URLs to inspect\n' +
  '\n' +
  'Example: clockdrift.js 15 http://tycho.usno.navy.mil/ http://www.example.com/\n'

const runSync = (args: readonly string[]): SpawnSyncReturns<string> => {
  return spawnSync('./clockdrift.js', args, { encoding: 'utf8' })
}

type runCallback = (code: string | null, error: Error | null) => void

const run = (args: readonly string[], cb: runCallback): ChildProcessWithoutNullStreams => {
  const subprocess = spawn('./clockdrift.js', args)
  subprocess.on('error', (err) => {
    assert.fail(err)
  })
  subprocess.on('exit', cb)
  return subprocess
}

{
  // Using no arguments should return the usage message.
  const ret = runSync([])
  assert.strictEqual(ret.stdout, '')
  assert.strictEqual(ret.stderr, usage)
  assert.strictEqual(ret.status, 1)
  assert.strictEqual(ret.signal, null)
}

{
  // The tolerance argument must be a number.
  const ret = runSync(['fhqwhgads', 'https://www.example.com/'])
  assert.strictEqual(ret.stdout, '')
  assert.strictEqual(ret.stderr, '\nError: tolerance must be a number.\n\n' + usage)
  assert.strictEqual(ret.status, 1)
  assert.strictEqual(ret.signal, null)
}

{
  // Invalid hostnames.
  const ret1 = runSync(['42', 'http://fhqwhgads.invalid/'])
  const ret2 = runSync(['42', 'https://fhqwhgads.invalid/'])
  ;
  [ret1, ret2].forEach((ret) => {
    assert.strictEqual(ret.stdout, '')
    assert.ok(ret.stderr.startsWith('Error on fhqwhgads.invalid: getaddrinfo '), ret.stderr)
    assert.strictEqual(ret.status, 1)
    assert.strictEqual(ret.signal, null)
  })
}

{
  // Unrecognized protocol.
  const ret = runSync(['1', 'fhqwhgads://'])
  assert.strictEqual(ret.stdout, '')
  assert.strictEqual(ret.stderr, 'Could not determine protocol: fhqwhgads:\n')
  assert.strictEqual(ret.status, 1)
  assert.strictEqual(ret.signal, null)
}

{
  // Start an http server on localhost and test the happy path.
  const server = http.createServer((req, res) => {
    res.writeHead(200, { Connection: 'close' })
    res.end('fhqwhgads')
  })
  server.listen(0, '127.0.0.1', () => {
    const addressInfo = server.address() as { port: number }
    const ret = run(
      ['1', `http://127.0.0.1:${addressInfo.port}/`],
      (code, signal) => {
        ret.stdout.on('data', assert.fail)
        ret.stderr.on('data', assert.fail)
        assert.strictEqual(code, 0)
        assert.strictEqual(signal, null)
        server.close()
      }
    )
  })
}

{
  // Test with http server that has an inaccurate Date header.
  const server = http.createServer((req, res) => {
    res.writeHead(200, {
      Date: new Date(Date.now() - 2000).toUTCString(),
      Connection: 'close'
    })
    res.end('fhqwhgads')
  })
  server.listen(0, '127.0.0.1', () => {
    const addressInfo = server.address() as { port: number }
    const port = addressInfo.port
    const ret = run(
      ['1', `http://127.0.0.1:${port}/`],
      (code, signal) => {
        ret.stderr.on(
          'data',
          (msg) => {
            const expected = `Clock at 127.0.0.1:${port} is -2s off from local clock.\n`
            assert.strictEqual(msg.toString(), expected)
          }
        )
        ret.stderr.on('data', assert.fail)
        assert.strictEqual(code, 0)
        assert.strictEqual(signal, null)
        server.close()
      }
    )
  })
}

{
  // Test with an http server that a malformed Date header.
  const server = http.createServer((req, res) => {
    res.setHeader('Date', 'fhqwhgads')
    res.setHeader('Connection', 'close')
    res.end('fhqwhgads')
  })
  server.listen(0, '127.0.0.1', () => {
    const addressInfo = server.address() as { port: number }
    const port = addressInfo.port
    const ret = run(
      ['1', `http://127.0.0.1:${port}/`],
      (code, signal) => {
        ret.stdout.on('data', assert.fail)
        ret.stderr.on(
          'data',
          (msg) => {
            const expected = `Could not convert date header from 127.0.0.1:${port} to timestamp. (Malformed?)\n`
            assert.strictEqual(msg.toString(), expected)
          }
        )
        assert.strictEqual(code, 1)
        assert.strictEqual(signal, null)
        server.close()
      }
    )
  })
}

{
  // Test with a broken http server that does not send a Date header.
  const server = http.createServer((req, res) => {
    res.sendDate = false
    res.writeHead(200, { Connection: 'close' })
    res.end('fhqwhgads')
  })
  server.listen(0, '127.0.0.1', () => {
    const addressInfo = server.address() as { port: number }
    const port = addressInfo.port
    const ret = run(
      ['1', `http://127.0.0.1:${port}/`],
      (code, signal) => {
        ret.stdout.on('data', assert.fail)
        ret.stderr.on(
          'data',
          (msg) => {
            const expected = `No date header returned by 127.0.0.1:${port}.\n`
            assert.strictEqual(msg.toString(), expected)
          }
        )
        assert.strictEqual(code, 1)
        assert.strictEqual(signal, null)
        server.close()
      }
    )
  })
}

{
  // Test with an http server that times out.
  const server = http.createServer()
  server.listen(0, '127.0.0.1', () => {
    const addressInfo = server.address() as { port: number }
    const port = addressInfo.port
    const msgs = [
      'Socket timeout; hanging up.',
      `Error on 127.0.0.1:${port}: socket hang up`
    ]
    const ret = run(
      ['1', `http://127.0.0.1:${port}/`],
      (code, signal) => {
        ret.stdout.on('data', assert.fail)
        ret.stderr.on(
          'data',
          (chunk) => {
            chunk.toString().split('\n').forEach((msg: string[]) => {
              if (msg.length > 0) {
                const expected = msgs.shift()
                assert.strictEqual(msg, expected)
              }
            })
          }
        )
        ret.stderr.on('end', () => {
          assert.deepStrictEqual(msgs, [])
        })
        assert.strictEqual(code, 1)
        assert.strictEqual(signal, null)
        server.close()
      }
    )
  })
}


clockdrift.ts
#!/usr/bin/env node

/*!
 * clockdrift.js by Rich Trott, (c) 2012 Regents of University of California, MIT License
 */

import http from 'node:http'
import https from 'node:https'
let i
const argvLength = process.argv.length

function usage (errorMessage: string): void {
  if (errorMessage !== '') {
    console.error(`\nError: ${errorMessage}.\n`)
  }
  console.error('Usage: clockdrift.js tolerance url1 [url2 ...]')
  console.error('    tolerance: number of seconds a timestamp can be off without being reported')
  console.error('    url1 ...: URLs to inspect')
  console.error('\nExample: clockdrift.js 15 http://tycho.usno.navy.mil/ http://www.example.com/')
}

function checkTime (host: string) {
  return function (res: http.IncomingMessage) {
    if (typeof res.headers.date !== 'string') {
      console.error(`No date header returned by ${host}.`)
      process.exitCode = 1
      return
    }

    const clockTimestamp = new Date(res.headers.date).getTime()
    if (isNaN(clockTimestamp)) {
      console.error('Could not convert date header from ' + host + ' to timestamp. (Malformed?)')
      process.exitCode = 1
      return
    }

    const diff = Math.round((clockTimestamp - before) / 1000)

    if (Math.abs(diff) > tolerance) {
      console.log(`Clock at ${host} is ${diff}s off from local clock.`)
    }
  }
}

function dispatchRequest (targetUrl: string): void {
  let reqObj
  let req: http.ClientRequest

  const options = new URL(targetUrl) as URL & { method: string }
  options.method = 'HEAD'

  switch (options.protocol) {
    case 'http:':
      reqObj = http
      break
    case 'https:':
      reqObj = https
      break
  }

  if (reqObj != null) {
    req = reqObj.request(options, checkTime(options.host))
    req.on('error', function (e) {
      console.error(`Error on ${options.host}: ${e.message}`)
      process.exitCode = 1
    })
    req.on('socket', function (socket) {
      socket.setTimeout(1000)
      socket.on('timeout', function () {
        console.error('Socket timeout; hanging up.')
        process.exitCode = 1
        req.abort()
      })
    })
    req.end()
  } else {
    console.error('Could not determine protocol: ' + options.protocol)
    process.exitCode = 1
  }
}

if (argvLength < 4) {
  usage('')
  process.exit(1)
}

const tolerance = parseInt(process.argv[2], 10)
if (isNaN(tolerance)) {
  usage('tolerance must be a number')
  process.exit(1)
}

const before = new Date().getTime()

for (i = 3; i < argvLength; i++) {
  dispatchRequest(process.argv[i])
}


clockdrift.js
#!/usr/bin/env node
/*!
 * clockdrift.js by Rich Trott, (c) 2012 Regents of University of California, MIT License
 */
import http from 'node:http';
import https from 'node:https';
let i;
const argvLength = process.argv.length;
function usage(errorMessage) {
    if (errorMessage !== '') {
        console.error(`\nError: ${errorMessage}.\n`);
    }
    console.error('Usage: clockdrift.js tolerance url1 [url2 ...]');
    console.error('    tolerance: number of seconds a timestamp can be off without being reported');
    console.error('    url1 ...: URLs to inspect');
    console.error('\nExample: clockdrift.js 15 http://tycho.usno.navy.mil/ http://www.example.com/');
}
function checkTime(host) {
    return function (res) {
        if (typeof res.headers.date !== 'string') {
            console.error(`No date header returned by ${host}.`);
            process.exitCode = 1;
            return;
        }
        const clockTimestamp = new Date(res.headers.date).getTime();
        if (isNaN(clockTimestamp)) {
            console.error('Could not convert date header from ' + host + ' to timestamp. (Malformed?)');
            process.exitCode = 1;
            return;
        }
        const diff = Math.round((clockTimestamp - before) / 1000);
        if (Math.abs(diff) > tolerance) {
            console.log(`Clock at ${host} is ${diff}s off from local clock.`);
        }
    };
}
function dispatchRequest(targetUrl) {
    let reqObj;
    let req;
    const options = new URL(targetUrl);
    options.method = 'HEAD';
    switch (options.protocol) {
        case 'http:':
            reqObj = http;
            break;
        case 'https:':
            reqObj = https;
            break;
    }
    if (reqObj != null) {
        req = reqObj.request(options, checkTime(options.host));
        req.on('error', function (e) {
            console.error(`Error on ${options.host}: ${e.message}`);
            process.exitCode = 1;
        });
        req.on('socket', function (socket) {
            socket.setTimeout(1000);
            socket.on('timeout', function () {
                console.error('Socket timeout; hanging up.');
                process.exitCode = 1;
                req.abort();
            });
        });
        req.end();
    }
    else {
        console.error('Could not determine protocol: ' + options.protocol);
        process.exitCode = 1;
    }
}
if (argvLength < 4) {
    usage('');
    process.exit(1);
}
const tolerance = parseInt(process.argv[2], 10);
if (isNaN(tolerance)) {
    usage('tolerance must be a number');
    process.exit(1);
}
const before = new Date().getTime();
for (i = 3; i < argvLength; i++) {
    dispatchRequest(process.argv[i]);
}
